### PR TITLE
Fix AWS IPPermission check for case with preexisting groups and ranges

### PR DESCRIPTION
**Pull request by therc**
If you have "alien" ranges (e.g. from peering), ipPermissionExists will never return true, because the length of the new range array will always be smaller than the existing one. So we'll end up trying to add a range that is already there, which causes EC2 to return a "duplicate rule" error and eventually throttle us (probably what might be happening in #20314). Turn the strict equality check into a subset check.

cc @justinsb

### CODE REVIEW

**Comment by davidopp** [comment]
@justinsb  can you take a look at this?

**Comment by therc** [comment]
Ping? I'm hoping I can get this in 1.2.

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 1]
```diff
@@ -832,6 +832,63 @@ func TestIpPermissionExistsHandlesMultipleGroupIds(t *testing.T) {
...
+		t.Errorf("Should have been considered existing since we're comparing a range array against itself")
+	}
+
+	exists = ipPermissionExists(&existingIpPermission, &oldIpPermission, false)
+	if !exists {
+		t.Errorf("Should have been considered existing since 10.* is in oldIpPermission's array of ranges")
+	}
+	exists = ipPermissionExists(&existingIpPermission2, &oldIpPermission, false)
+	if !exists {
+		t.Errorf("Should have been considered existing since 172.* is in oldIpPermission2's array of ranges")
```
Comment (davidopp) [inline pkg/cloudprovider/providers/aws/aws_test.go]: this should say 192.* I think?

Comment (therc) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Good catch. I fixed that and a misspelling of 'against'.

**Comment by davidopp** [comment]
LGTM

**Comment by justinsb** [comment]
Sorry @therc - can you explain a bit why this was needed?

ipPermissionExists is called from ensureSecurityGroupIngress and removeSecurityGroupIngress.

In removeSecurityGroupIngress it seems incorrect.  We will remove a security group ingress rule that is potentially wider than what we intended.

ensureSecurityGroupIngress is called as part of the LoadBalancer creation/update code.  It is called for the ELB security group and the instance security group.  Both groups are owned by k8s.  In neither case should any peered IPs be present.

I am planning on fixing this as part of the fix for #21895, which appears to be a long-standing issue but I only noticed as part of #21907.  The fix will set the ELB security groups to be exactly what the user has requested them to be - no extra rules allowed unless specified through k8s using the new security group annotation.

**Comment by therc** [comment]
This happens when you run in the default VPC and the cloud provider picks the first security group for the instance, I think. There's no guarantee that we own the latter.

**Comment by justinsb** [comment]
Ah OK.  We should make sure that security groups are tagged, and that the cloudprovider checks the tags.  For backcompat I guess we'll have to only check the tags if we find more than one security group.

I've opened #21986, so we can be sure that we can rely on tagging.  The k8s-owned security group will be owned by k8s.  If you want to add another security group, that's OK as long as it isn't tagged :-)

**Pull request merged**